cluster.go

```diff
@@ -459,13 +475,79 @@ func (c *Cluster) Do(a Action) error {
 	return c.doInner(a, addr, key, false, doAttempts)
 }
 
+func (c *Cluster) getClusterDownSince() time.Time {
+	t, _ := c.lastClusterdown.Load().(time.Time)
+	return t
+}
+
+func (c *Cluster) setClusterDown(down bool) (changed bool) {
+	// avoid taking the lock when the down state was already cleared
+	prevVal, _ := c.lastClusterdown.Load().(time.Time)
+	if !down && prevVal.IsZero() {
+		return false
+	}
+
+	var newVal time.Time
+	if down {
+		// check for a small difference in time to reduce contention
+		if newVal = time.Now(); newVal.Sub(prevVal) < time.Millisecond {
```

Comment:
If in the future lock contention becomes to noticeable we could increase the duration here (or even try to be smart with the configured sleep duration), but I think this should work well enough for now.